app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt

```diff
@@ -331,6 +335,11 @@ class SessionControlAdapter(
                 // This View already listens and maps store updates. Avoid creating and binding new Views.
                 // The composition will live until the ViewTreeLifecycleOwner to which it's attached to is destroyed.
             }
+            is CollectionViewHolder -> {
+                // Dispose the underlying composition immediately.
+                // This ViewHolder can be removed / re-added and we need it to show a fresh new composition.
+                holder.composeView.disposeComposition()
```

Comment:
Good question.
Right now all composable viewholders will set a different than default strategy [here](https://github.com/mozilla-mobile/fenix/blob/58a6fafabc0477f6fa8f56adab8d37f849869e52/app/src/main/java/org/mozilla/fenix/compose/ComposeViewHolder.kt#L35-L37).
[DisposeOnViewTreeLifecycleDestroyed](https://developer.android.com/reference/kotlin/androidx/compose/ui/platform/ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed) rather than the default [DisposeOnDetachedFromWindow](https://developer.android.com/reference/kotlin/androidx/compose/ui/platform/ViewCompositionStrategy.DisposeOnDetachedFromWindow).

For them being inside a RecyclerView that may recycle holders the default would mean that when the holders are offscreen they would get recycled and the composition disposed 
to afterwards having to be recreate it when we'd need to show the same homescreen section again.
So setting `DisposeOnViewTreeLifecycleDestroyed` was a performance improvement. Create the holder, the composition and do not dispose it while the user is on the homescreen, only pay the penalty of creating these once.

This worked great for composable viewholders which would observe the store and update themselves.
In the case of the collections viewholders the update flow is a bit more convoluted. They will not observe the store but have the state injected.
In this case I had to go back to the default - `DisposeOnDetachedFromWindow` strategy to avoid them being still being shown on the screen after being deleted to prevent state leakage - the old holders showing stale data - what they previously displayed. This can be seen by commenting that line and deleting the collection / collection items and then undoing that action.
Keeping the default `DisposeOnDetachedFromWindow` in these cases means basically the same flow as for the xml version and ties nicely with the current way of passing data to these holders (highly convoluted, will probably need to be refactored later).
`CollectionViewHolder` could do something similar to [`ComposeViewHolder`](https://github.com/mozilla-mobile/fenix/blob/58a6fafabc0477f6fa8f56adab8d37f849869e52/app/src/main/java/org/mozilla/fenix/compose/ComposeViewHolder.kt#L35-L37) and override the strategy in it's init so we'd not have to do this `disposeComposition()` here but that would mean the holder knows where and how it is used, which it shouldn't, so I kept the default for it but managed the composition inside the adapter which is the actual component responsable for deciding what gets shown.

=> This approach is useful when state is passed from outside to avoid state leakage. 
We only need it because the compose strategy was overriden to be `DisposeOnViewTreeLifecycleDestroyed` and so holders will still remain on the screen until the screen gets destroyed.

Behavior without this:

https://user-images.githubusercontent.com/11428869/167076322-cd42dcd3-bae1-4396-9ffd-ce604e399fc7.mp4